zkevm-circuits/src/evm_circuit/execution/exp.rs

```diff
@@ -0,0 +1,304 @@
+            from_bytes::expr(&exponent_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponent_rlc.cells[0x10..0x20]),
+        );
+        // Extract low and high bytes of the exponentiation result.
+        let (exponentiation_lo, exponentiation_hi) = (
+            from_bytes::expr(&exponentiation_rlc.cells[0x00..0x10]),
+            from_bytes::expr(&exponentiation_rlc.cells[0x10..0x20]),
+        );
+
+        // We simplify constraints depending on whether or not the exponent is 0 or 1.
+        // In order to do this, we build some utility expressions.
+        let exponent_lo_is_zero = IsZeroGadget::construct(cb, exponent_lo.clone());
+        let exponent_hi_is_zero = IsZeroGadget::construct(cb, exponent_hi.clone());
+        let exponent_is_zero_expr =
+            and::expr([exponent_lo_is_zero.expr(), exponent_hi_is_zero.expr()]);
+        let exponent_lo_is_one = IsEqualGadget::construct(cb, exponent_lo.clone(), 1.expr());
+        let exponent_is_one_expr =
+            and::expr([exponent_lo_is_one.expr(), exponent_hi_is_zero.expr()]);
+
+        let zero_rlc = cb.query_word();
+        cb.require_zero(
+            "base * base + c == base^2 (c == 0)",
+            sum::expr(&zero_rlc.cells),
+        );
+        let base_sq = cb.query_word();
+
+        // If exponent == 0, base^exponent == 1, which implies:
+        // 1. Low bytes of exponentiation == 1
+        // 2. High bytes of exponentiation == 0
+        cb.condition(exponent_is_zero_expr.clone(), |cb| {
+            cb.require_equal(
+                "exponentiation == 1 if exponent == 0 (lo == 1)",
+                exponentiation_lo.clone(),
+                1.expr(),
+            );
+            cb.require_equal(
+                "exponentiation == 1 if exponent == 0 (hi == 0)",
+                exponentiation_hi.clone(),
+                0.expr(),
+            );
+        });
+        // If exponent == 1, base^exponent == base, which implies:
+        // 1. Low bytes of exponentiation == low bytes of base.
+        // 2. High bytes of exponentiation == high bytes of base.
+        cb.condition(exponent_is_one_expr.clone(), |cb| {
+            cb.require_equal(
+                "exponentiation == base if exponent == 1 (lo)",
+                exponentiation_lo.clone(),
+                base_lo.clone(),
+            );
+            cb.require_equal(
+                "exponentiation == base if exponent == 1 (hi)",
+                exponentiation_hi.clone(),
+                base_hi.clone(),
+            );
+        });
+        // If exponent > 1, i.e. exponent != 0 && exponent != 1:
+        // We do two lookups to the exponentiation table. If exponent == 2, there is
+        // only a single step in the exponentiation by squaring trace. In this
+        // case, is_first == is_last == true for that step.
+        let single_step = cb.query_cell();
+        cb.condition(
+            not::expr(or::expr([exponent_is_zero_expr, exponent_is_one_expr])),
```

Comment:
The `or::expr` is implemented with `not (not A and not B)`, so `not or` would become `not not (not A and not B)`. I chatted with @han0110 offline, and the conclusion is that
- it wouldn't be a verifier side cost because the expression is evaluated.
- it might be a prover side cost. Although @Brechtpd had some optimizations here https://github.com/privacy-scaling-explorations/halo2/blob/main/halo2_proofs/src/plonk/evaluation.rs#L589-L690, the expression might incur the `2^extended_k` cost in the prover side. It would be preferable to use a simplified one on the circuit side.

```suggestion
            and::expr([not::expr(exponent_is_zero_expr), not::expr(exponent_is_one_expr)]),
```

Reply:
This is interesting. I hadn't checked the implementation of the `or::expr` before. I'll make this change :)